code/game/machinery/shieldgen.dm

```diff
@@ -553,3 +553,644 @@
+	innate_radius = 3
+	innate_strength = 40
+
+	for(var/datum/stock_part/capacitor/new_capacitor in component_parts)
+		innate_strength += new_capacitor.tier * 10
+
+	for(var/datum/stock_part/servo/new_servo in component_parts)
+		innate_regen += new_servo.tier
+
+	for(var/datum/stock_part/micro_laser/new_laser in component_parts)
+		innate_radius += new_laser.tier * 0.25
+
+	calculate_regeneration()
+	calculate_max_strength()
+	calculate_radius()
+
+
+/obj/machinery/modular_shield_gen/Initialize(mapload)
+	. = ..()
+	deployed_shields = list()
+	connected_modules = list()
+	list_of_turfs = list()
+	wires = new /datum/wires/modular_shield_gen(src)
+	if(mapload && active && anchored)
+		activate_shields()
+
+/datum/wires/modular_shield_gen
+	proper_name = "Modular shield generator"
+	randomize = FALSE
+	holder_type = /obj/machinery/modular_shield_gen
+
+/datum/wires/modular_shield_gen/New(atom/holder)
+	wires = list(WIRE_HACK)
+	..()
+
+/datum/wires/modular_shield_gen/on_pulse(wire)
+
+	var/obj/machinery/modular_shield_gen/G = holder
+	switch(wire)
+		if(WIRE_HACK)
+			G.toggle_shields()
+			return
+	..()
+
+/obj/machinery/modular_shield_gen/proc/deactivate_shields()
+	active = FALSE
+	QDEL_LIST(deployed_shields)
+	calculate_regeneration()
+
+/obj/machinery/modular_shield_gen/attackby(obj/item/W, mob/user, params)
+
+	if(default_deconstruction_screwdriver(user,"gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"]open",
+		"gen_[!(machine_stat & NOPOWER) ? "[recovering ? "recovering_" : "ready_"]" : "no_power_"]closed",  W))
+		return
+
+	if(default_deconstruction_crowbar(W) && !(active) && !(recovering))
+		return
+
+	if(is_wire_tool(W) && panel_open)
+		wires.interact(user)
+		return
+
+	return ..()
+
+/obj/machinery/modular_shield_gen/proc/toggle_shields()
+	if(active)
+		deactivate_shields()
+		return
+	if (recovering)
+		return
+	activate_shields()
+
+
+
+/obj/machinery/modular_shield_gen/proc/activate_shields()
+	if(active) //bug or did admin call proc on already active shield gen?
+		return
+	active = TRUE
+
+	if(radius >= 10) //the shield is large so we are going to use the midpoint formula and clamp it to the lowest full number in order to save processing power
+		var/fradius = round(radius)
+		var/list/inside_shield = circle_range_turfs(src, fradius - 1)//in the future we might want to apply an effect to turfs inside the shield
+		var/t1 = fradius/16
+		var/dx = fradius
+		var/dy = 0
+		var/t2
+		var/list/list_of_turfs = list()
+		while(dx >= dy)
+			list_of_turfs += locate(x + dx, y + dy, z)
+			list_of_turfs += locate(x - dx, y + dy, z)
+			list_of_turfs += locate(x + dx, y - dy, z)
+			list_of_turfs += locate(x - dx, y - dy, z)
+			list_of_turfs += locate(x + dy, y + dx, z)
+			list_of_turfs += locate(x - dy, y + dx, z)
+			list_of_turfs += locate(x + dy, y - dx, z)
+			list_of_turfs += locate(x - dy, y - dx, z)
+			dy += 1
+			t1 += dy
+			t2 = t1 - dx
+			if(t2 > 0)
+				t1 = t2
+				dx -= 1
```

Comment:
what the fuck? look for stuff in places like `/code/_HELPERS/maths.dm` for procs that do stuff like this for you already, there has to be something like `get_angle` or whatever you need for this. if there isn't something already, i would rather you add this part of the code to `maths.dm` so it can at least be used to help other people hsould they really truly need it

Reply:
there is a reason why i made this only run when the shield radius gets very large, it generates a pretty ugly circle when its smaller and it takes decimals very horribly badly.

ill look into that though.

Reply:
you can have it read the radius and then alternate between two procs